### PR TITLE
Classification Prediction

### PR DESCRIPTION
**Pull request by Sllambias**
Okay this turned out to be a pretty large update.
I've tested that it runs with
Classification: preproc, train, inference
Segmentation: Preproc, train, inference
Unsupervised: Preproc, train

What are we getting:
 
TLDR:

Big changes:
- [x] All Preprocessors now rely on the same functions at core instead of having clones of the same functions with very minor differences. This is the brunt of the PR and includes a lot of rewriting/cleaning of the Preprocessors.
- [x] YuccaNet now has predict method and sliding_window_predict (2D/3D) methods. This supports Classification prediction, full-size image predictions and sliding window predictions (patch-based)
- [x] Planners take arguments for fixed size to enable models not using patch-based training. Additionally it can be specified if aspect ratio must be kept or not. This means EITHER target_spacing OR target_size can be specified in planner (or left to be inferred based on the median spacing as the default)
- [x] Manager supports training on full images (toggled by patch_based_training)
- [x] Inference supports prediction on full-size images for models trained without patch_based_training

Minor fixes
- [x] Fixed small issue where inference failed if the preprocessing folder was missing for a model, because the splits file could not be found/created. This is not desired as the splits are not needed in inference
- [x] Improved padding functionality to take more args and have better defaults
- [x] Small corrections with incorrect variable names 
- [x] Edited some run_X scripts to include all the new args

### CODE REVIEW

**Comment by asbjrnmunk** [comment]
Since this is mostly (for now) to be used by @jakobamb, I'll wait to give my review until after him :-)

File: yucca/training/configuration/configure_input_dims.py [hunk 1]
```diff
@@ -21,14 +21,21 @@ def lm_hparams(self):
...
     batch_size: Union[int, Literal["tiny"]] = None,
+    patch_based_training: bool = True,
     patch_size: Union[Literal["max", "min", "mean", "tiny"], Tuple[int, int], Tuple[int, int, int], None] = None,
 ):
     num_modalities = max(1, plan.get("num_modalities") or len(plan["dataset_properties"]["modalities"]))
 
+    if patch_based_training is False:
+        assert plan.get("new_max_size") == plan.get(
+            "new_min_size"
+        ), "sizes in dataset are not uniform. Patch based training only works for datasets with uniform data shapes."
```
Comment (jakobamb) [inline yucca/training/configuration/configure_input_dims.py:36]: needs to be changed to "Non-patch based training only works for datasets with uniform data shapes."

**Comment by Sllambias** [comment]
@asbjrnmunk Jakob only had minor comments (which I all agree with), so once you've also reviewed I will start fixing things

File: yucca/network_architectures/networks/YuccaNet.py [hunk 2]
```diff
@@ -30,14 +30,16 @@ def load_state_dict(self, target_state_dict, *args, **kwargs):
...
-        if mode == "3D":
-            predict = self._predict3D
-        if mode == "2D":
-            predict = self._predict2D
+        if not sliding_window_prediction:
+            return self._predict(data)
+        elif mode == "3D":
+            predict = self._sliding_window_predict3D
+        elif mode == "2D":
+            predict = self._sliding_window_predict2D
```
Comment (asbjrnmunk) [inline yucca/network_architectures/networks/YuccaNet.py:42]: While python allows for otherwise, i find it best that a branch always returns values of the same type. The first branch returns a value, while the two other branches returns methods. I would prefer if we rewrote it to:
```python
if not sliding_window_prediction:
    return self._predict(data)
            
if mode == "3D":
    predict = self._sliding_window_predict3D
elif mode == "2D":
     predict = self._sliding_window_predict2D
 ...
```

File: yucca/network_architectures/networks/YuccaNet.py [hunk 3]
```diff
@@ -57,7 +59,15 @@ def predict(self, mode, data, patch_size, overlap, mirror=False):
             pred /= div
         return pred
 
-    def _predict3D(self, data, patch_size, overlap):
+    def _predict(self, data):
```
Comment (asbjrnmunk) [inline yucca/network_architectures/networks/YuccaNet.py:62]: `YuccaNet` now contains two methods `_predict` and `predict` but the two methods are quite different. I would prefer if one of the two methods had names which made it clear how they are different. My guess would be to rename `_predict` to something like `_full_size_predict` or `_standard_predict` or similar.

File: yucca/planning/YuccaPlanner.py [hunk 4]
```diff
@@ -105,8 +117,16 @@ def determine_transpose(self):
...
-    def determine_spacing(self):
-        self.target_spacing = np.median(self.dataset_properties["original_spacings"], 0).tolist()
+    def determine_target_spacing_or_size(self):
+        self._determine_target_spacing_or_size()
+        assert self.target_size is None or self.target_spacing is None, (
+            "only one of target size or target spacing should be specified"
+            f" but both are specified here as {self.target_size} and {self.target_spacing} respectively"
+        )
+
+    def _determine_target_spacing_or_size(self):
```
Comment (asbjrnmunk) [inline yucca/planning/YuccaPlanner.py:127]: Why do we have to have both a private and a public `determine_target_spacing_or_size`? I dont think it would be bad if we just added the logic of `_determine_target_spacing_or_size` into `determine_target_spacing_or_size`.

File: yucca/training/configuration/split_data.py [hunk 5]
```diff
@@ -40,6 +41,9 @@ def get_split_config(
         - If `k` is provided we will split with `k-fold`.
         - If `p` is provided it determines the fraction of items used for the val split.
     """
+    if stage == "predict":
```
Comment (asbjrnmunk) [inline yucca/training/configuration/split_data.py:44]: I would much rather have that we just change the manager such that it branches in the split_config than make a [nop](https://en.wikipedia.org/wiki/NOP_(code)) based on an argument value. From the caller side i find this very hard to understand this code. I would do this instead in `YuccaManager`:

```python
splits_config = get_split_config(train_data_dir=path_config.train_data_dir, task=task_config.task) if stage != 'predict' else SplitConfig()
```

File: yucca/training/managers/alternative_trainers/YuccaManager_NoPatches.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# %%
+from yucca.training.managers.YuccaManager import YuccaManager
+
+
+class YuccaManager_NoPatches(YuccaManager):
```
Comment (asbjrnmunk) [inline yucca/training/managers/alternative_trainers/YuccaManager_NoPatches.py:5]: Just to be frank, I find this to be an abusive use of inheritance :-) I think this should be a parameter, however understand that it would screw up the `save_path`, (while we should fix that some other way....). But not sure now is the time to fix it.

tumblr_nb6izd25Ao1tkl4vco1_500

**Review by asbjrnmunk — COMMENTED** [review]
Really cool update! I love that you are deleting _almost_ as many lines as you add (off by 5!). I have a few comments, where one is big, but shouldn't necessarily block this PR.

File: yucca/planning/YuccaPlanner.py (hunk 4 above)
Comment (Sllambias) [reply yucca/planning/YuccaPlanner.py:127]: So the reason for this was that you could then overwrite the private version without overwriting the validation check, but Jakob also pointed this out and I ended up making just 1 public version and an additional function (with completely different name) that validates the variables are fine

**Comment by Sllambias** [comment]
@asbjrnmunk @jakobamb Thanks for the comments! I think we're more or less ready to merge.
@asbjrnmunk all your comments are fixed now (except the one about my depraved use of Classes, that's just who I am 👼 ). But which of your comments was the big one you mentioned? I found the suggested changes to be not too large and pretty straight forward to implement

**Comment by Sllambias** [comment]
Ah yes, pushing. Very good idea.

**Review by asbjrnmunk — APPROVED** [review]
